### ISSUE TITLE
Merge function in SimpleTruthValue.cc does not implement the right math

### ISSUE THREAD

**Issue by bgoertzel** [opencog/opencog#1279]
Currently the merge function for SimpleTruthValue appears to merge old and new TVs via choosing the one with the highest confidence.   This is not a good formula.   A host of better, simple formulas for this are given in the PLN book.   See e.g. section 5.10.2 where the formula for merging (s1, n1) with (s2,n2) is given as:

s = s1 w1 + s2 w2, where w1 = n1 /(n1 +n2), w2 = n2 / (n1+n2)

n = n1 + n2 - c min(n1,n2)

c is a parameter between 0 and 1 ....  Ideally there would be a version of merge taking c as a parameter, and then a version using a default value for c (which could be c=0 or c=.2 or whatever...)

The STV uses (s,c) where c = n /(n+k), where n is the count...

Getting this merge function right is important because without it we can't use SimpleTruthValue to accumulate frequency information about data.

**Comment by linas** [comment]
truth-value merge is a sticky problem in the atomspace.

1) there is no way to selected a "prefered" truth-value merge formula (as different things may want different formulas)

2) there is no clear way for merging truth values from two different cogservers that are sharing data.  Lets say they are sharing via the sql backend. So cogserver A writes out a TV,  cogserver B pulls that value, merges with its own private copy, writes out the merged result.  cogserver A notices the TV changed, pulls that TV, merges with its own TV, writes out the result.  Now repeat ... 

3) confusion about merging default TV's. was the default TV there "by accident", and thus should not be merged?  Or did we really mean to average it into the mix?

4) if there are multiple different kinds of truth values, how should these be merged?

5) When is a SimpleTV a "Fuzzy set membership TV" and when is it something else?

These kinds of questions made me conclude that "automatic" merging of TV's in the atomspace is almost surely a bad idea, and that the algo that is generating the TV's should take explicit control, and decide what to do in each case, so that its done right for that particular ago.

Regarding frequency: keeping track of frequency is on reason I introduced the "CountTV",  with a third field devoted exclusively to counting.  Since that time, I also have a burning desire for a fourth field: a place to hold the (logarithm of) the frequency (i.e. a normalized count).  I have no thoughts about how these last two could/should interact with the traditional SimplTV strength/confidence values.

**Comment by bgoertzel** [comment]
I think that CountTV and SimpleTV are very similar...  a CountTV can always
be interpreted as a SimpleTV  straighforwardly; and a SimpleTV can always
be cast into a CountTV heuristically (based on making some
assumptions)....  So merging a CountTV and a SimpleTV, on a pragmatic
basis, can be done with no issues.

For instance, in practice if one converts using the formula

confidence = count / (count + k)

where k is a parameter, one will get something generally sensible...

Whether the merger of a CountTV and a SimpleTV should yield a CountTV or a
SimpleTV really depends on what you're doing.....

It seems to me there SHOULD be a TV merge that just merges two TV objects
in a reasonable way, but also a fancier merge that lets you choose what
kind of TV should be produced as an output.   A simple way to do this would
be to have a merge method with three options.  When merging new TV tv_new
with existing TV (in the Atomspace) tv_old, the options are

-- take the TV type of tv_old
-- take the TV type of tv_new
-- take the TV type that the merge method decides you should take

I.e. the method would look conceptually like

TruthValue.merge(TruthValue tv_new, int tvMergeStrategy)

where the code tvMergeStrategy lets you choose which of the 3 above merge
strategies to use.   The default

TruthValue.merge(TruthValue tv_new)

would assume the option "take the TV type that the merge method decides you
should take"

Something like that would serve for the time being I think...

Creators of AI code should have the option to do merging themselves in
their own fancy way.  However, having default methods for doing merging
also seems critical...

Currently, the default merge method for SimpleTV is too simple and "wrong";
the revised version I've suggested will work better....

I would say we also need a default method for merging SimpleTV and CountTV
since both are in active use at the moment.

-- Ben

On Thu, Feb 19, 2015 at 4:36 AM, Linas Vepštas <notifications@github.com>
wrote:

> truth-value merge is a sticky problem in the atomspace.
>
> 1) there is no way to selected a "prefered" truth-value merge formula (as
> different things may want different formulas)
>
> 2) there is no clear way for merging truth values from two different
> cogservers that are sharing data. Lets say they are sharing via the sql
> backend. So cogserver A writes out a TV, cogserver B pulls that value,
> merges with its own private copy, writes out the merged result. cogserver A
> notices the TV changed, pulls that TV, merges with its own TV, writes out
> the result. Now repeat ...
>
> 3) confusion about merging default TV's. was the default TV there "by
> accident", and thus should not be merged? Or did we really mean to average
> it into the mix?
>
> 4) if there are multiple different kinds of truth values, how should these
> be merged?
>
> 5) When is a SimpleTV a "Fuzzy set membership TV" and when is it something
> else?
>
> These kinds of questions made me conclude that "automatic" merging of TV's
> in the atomspace is almost surely a bad idea, and that the algo that is
> generating the TV's should take explicit control, and decide what to do in
> each case, so that its done right for that particular ago.
>
> Regarding frequency: keeping track of frequency is on reason I introduced
> the "CountTV", with a third field devoted exclusively to counting. Since
> that time, I also have a burning desire for a fourth field: a place to hold
> the (logarithm of) the frequency (i.e. a normalized count). I have no
> thoughts about how these last two could/should interact with the
> traditional SimplTV strength/confidence values.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/issues/1279#issuecomment-74942803>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by linas** [comment]
I have no complaints about changing the default; I don't care (the default is more-or-less unchanged from 10 years ago).   I was just saying that there are other issues as well.  A year or two ago, you pointed out that some atoms treat TV's as fuzzy set membership, while other atoms treat TV's as bayesian probabilities.  I presume that, because they are different, that the merge formulas would be different, too. I never quite figured out which atoms are supposed to use which. 

Also:
a) when you merge a simple and a count, what do you get when you are done, a simple, or a count?
b) how do you merge indefinite TV's?  What happens when other TV types are merged into indefinite, and vice-versa?

I don't have any clear idea of the "right way" to handle all of these different questions.   Just saying that, as a practical matter they come up as soon as one gets into the code (the old code for this was a complete spaghetti code jungle, what remains now is the simplified version that I teased out of it.)

**Comment by bgoertzel** [comment]
***
A year or two ago, you pointed out that some atoms treat TV's as fuzzy set
membership, while other atoms treat TV's as bayesian probabilities. I
presume that, because they are different, that the merge formulas would be
different, too.
***


Well, the thing is that fuzzy math doesn't really come with any nice way to
handle belief revision and weights of evidence and so forth.  There are
lots of related papers  in the academic literature, but nothing
particularly compelling or simple.  So in practice it seems OK to me to
simply use the probabilistic formulas for belief revision (merging) for
cases where SimpleTV represents a fuzzy value as well....  At least I
didn't see a case where this would cause an actual problem...

So we could create a SimpleProbabilisticTV and SimpleFuzzyTV, and this
would probably be a good idea just for conceptual clarity.   Then
EvaluationLinks would get SimpleFuzzyTV, whereas InheritanceLinks e.g.
would get SimpleProbabilisticTV....  However, as an initial default we'd
use the same merger rules for the two kinds of SimpleTV anyway....  So this
would be a change for conceptual transparency of the code.  Whereas
changing the merge rule for SimpleTV is a matter of urgent necessity, as
the current merge rule will cause wrong behaviors....



Also:
> a) when you merge a simple and a count, what do you get when you are done,
> a simple, or a count?
>


That depends on what you want to do.  The simple answer is, it doesn't make
much difference.  I suggested in my last email to allow either option to
the user.


> b) how do you merge indefinite TV's? What happens when other TV types are
> merged into indefinite, and vice-versa?
>

My suggestion for a default is, when merging a complex TV with a simple
one, to have the result be a simple one.  In the case of IndefiniteTV and
SimpleTV/CountTV, the easiest thing is to have the merged result be a
SimpleTV/CountTV...


> I don't have any clear idea of the "right way" to handle all of these
> different questions. Just saying that, as a practical matter they come up
> as soon as one gets into the code (the old code for this was a complete
> spaghetti code jungle, what remains now is the simplified version that I
> teased out of it.)
>

mathematically/conceptually these issues have been thought through a lot
years ago, and are all pretty simple ....  But we haven't bothered to make
the code reflect our thoughts on the matter....  Now seems a good time...

ben

**Comment by linas** [comment]
>
>
> mathematically/conceptually these issues have been thought through a lot
> years ago, and are all pretty simple ....
>

Ah Ben, you are infuriating.  You carefully and cleverly ducked every last
issue that I raised, so these all remain open issues. They may not be
urgent, but they remain unresolved. So don't say its "conceptually simple"
because there is no clarity whatsoever in this.   Its not even vaguely
close to being conceptually simple. If it was, it would have been done
right eons ago.

You can change the default formulas to be anything you want -- none of my
code depends on them, since I long ago realized its a snake pit. It will
remain a snake pit until a solution is developed.  Nothing you write above
suggests a solution.  Just sayin'.

**Comment by bgoertzel** [comment]
Hmmm. I don't think I ducked the issues... the thing is that the specifics
of the issues (regarding the SimpleTV and IndefiniteTV at any rate) are
already addressed in the PLN book, and I didn't paste the details into my
email....  It's not ducking, it's just deferring to a published reference ;)

Anyway  Matt Ikle' is full-time on OpenCog thru June (visiting HK), so he
and I will fix all this stuff in the fairly near future, including
extensive code comments that hopefully won't leave you feeling anything is
being ducked...

BTW the text of all my books on OpenCog-related details is going to end up
on the wiki site sometime in the next few months, which should move some
distance toward solving the "opaque conceptual details" problem...

ben

On Thu, Feb 19, 2015 at 12:21 PM, Linas Vepštas <notifications@github.com>
wrote:

> >
> >
> > mathematically/conceptually these issues have been thought through a lot
> > years ago, and are all pretty simple ....
> >
>
> Ah Ben, you are infuriating. You carefully and cleverly ducked every last
> issue that I raised, so these all remain open issues. They may not be
> urgent, but they remain unresolved. So don't say its "conceptually simple"
> because there is no clarity whatsoever in this. Its not even vaguely
> close to being conceptually simple. If it was, it would have been done
> right eons ago.
>
> You can change the default formulas to be anything you want -- none of my
> code depends on them, since I long ago realized its a snake pit. It will
> remain a snake pit until a solution is developed. Nothing you write above
> suggests a solution. Just sayin'.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/issues/1279#issuecomment-74997204>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by bgoertzel** [comment]
How to tweak the existing code to solve the problem, without impacting any
of your current work with CountTV or messing up anything else currently
being done, seems clear to me -- but it will be easier for me just to
change the code, or to sit next to Matt/Misgana while we pair or triple
program the simple changes to the code, than to type about it endlessly in
comments...

On Thu, Feb 19, 2015 at 12:33 PM, Ben Goertzel <ben@goertzel.org> wrote:

>
> Hmmm. I don't think I ducked the issues... the thing is that the specifics
> of the issues (regarding the SimpleTV and IndefiniteTV at any rate) are
> already addressed in the PLN book, and I didn't paste the details into my
> email....  It's not ducking, it's just deferring to a published reference ;)
>
> Anyway  Matt Ikle' is full-time on OpenCog thru June (visiting HK), so he
> and I will fix all this stuff in the fairly near future, including
> extensive code comments that hopefully won't leave you feeling anything is
> being ducked...
>
> BTW the text of all my books on OpenCog-related details is going to end up
> on the wiki site sometime in the next few months, which should move some
> distance toward solving the "opaque conceptual details" problem...
>
> ben
>
> On Thu, Feb 19, 2015 at 12:21 PM, Linas Vepštas <notifications@github.com>
> wrote:
>
>> >
>> >
>> > mathematically/conceptually these issues have been thought through a lot
>> > years ago, and are all pretty simple ....
>> >
>>
>> Ah Ben, you are infuriating. You carefully and cleverly ducked every last
>> issue that I raised, so these all remain open issues. They may not be
>> urgent, but they remain unresolved. So don't say its "conceptually simple"
>> because there is no clarity whatsoever in this. Its not even vaguely
>> close to being conceptually simple. If it was, it would have been done
>> right eons ago.
>>
>> You can change the default formulas to be anything you want -- none of my
>> code depends on them, since I long ago realized its a snake pit. It will
>> remain a snake pit until a solution is developed. Nothing you write above
>> suggests a solution. Just sayin'.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/opencog/opencog/issues/1279#issuecomment-74997204>.
>>
>
>
>
> --
> Ben Goertzel, PhD
> http://goertzel.org
>
> "The reasonable man adapts himself to the world: the unreasonable one
> persists in trying to adapt the world to himself. Therefore all progress
> depends on the unreasonable man." -- George Bernard Shaw
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by linas** [comment]
I think we are talking past each other.  You are talking about specific formulas. But that's not the issue that I'm talking about.  As I said before, you are welcome to code up any formula what-so-ever.  Mostly I couldn't care, because the specific numerical values and formulas used are irrelevant to the issue of how truth values are handled, from the engineering perspective.  That is not where the snakes lie.

To change the formula for the simple truth value, just go here:
https://github.com/opencog/opencog/blob/master/opencog/atomspace/SimpleTruthValue.cc#L75
and code up any formula desired.  Changing the formula there does not in any way alter any of the issues that I raised.  It has no bearing on any of them.

**Comment by bgoertzel** [comment]
Linas,

Yeah, there are various issues regarding merging of the various TV types

A couple are:

1)
Having the right merge formula for merger of one TV of type X, with another
TV of type X.   As you say, this is just a matter of putting the right
formula into the code.  Not really an engineering decision, just a matter
of getting the math/concepts right

2)
What do we do when merging a TV of type X into a TV of type Y?    This is
hairier...

I think it should be fair to assume, for the time being, that two TV types
being assigned to the same Atom type, should be mutually interconvertible
(perhaps with some loss of information).   If we assume this, then when
merging a TV tv1 of type X into a TV tv2 of type Y, we can assume there
exist methods to cast tv1 into type Y, or to cast tv2 into type X

So, suppose we have an Atom A with a TV tv1 of type X.  Suppose we want to
merge a new TV tv2 of type Y into A's truth value....  So we are calling
some method like

A.merge(tv2)

One way to look at this is

A.merge(tv2, typeoption = old | new | default | stronger | weaker)

If typeoption=old, then the type of the resulting TV is tv1

If typeoption=new, then the type of the resulting TV is tv2

If typeoption=dominant, then the type of the resulting TV is whatever is
specified in the merge code as the default for merging a TV of type X and a
TV of type Y (this is going to be either X or Y)

For instance, for merging a SimpleTV with an IndefiniteTV, the default
might be IndefiniteTV

If typeoption=stronger then the type of the resulting TV is whichever of X
or Y is specified as stronger

If typeoption=weaker then the type of the resulting TV is whichever of X or
Y is specified as weaker

What I mean by "stronger vs. weaker" here is e.g., SimpleProbabilisticTV is
weaker than IndefiniteTV, because an IndefiniteTV can exactly specify a
SimpleProbabilisticTV, but a Simple ProbabilisticTV only specified an
IndefiniteTV with a lot of heuristic fudging.

...

This kind of approach would require specifying somewhere or other, the
strength ordering and default ordering of each pair of commensurable TV
types (where "commensurable" means that the two TV types are applicable to
the same Atom types)...

It seems to me this would give enough flexibility for different cognitive
processes to do what they wish with TV types....   And if someone doesn't
want to worry about it, they can call

A.merge(tv2) = A.merge(tv2, typeoption=default)

and then magic will happen and hopefully the default choices won't cause
any egregious stupidity ;p

-- Ben









On Thu, Feb 19, 2015 at 1:07 PM, Linas Vepštas <notifications@github.com>
wrote:

> I think we are talking past each other. You are talking about specific
> formulas. But that's not the issue that I'm talking about. As I said
> before, you are welcome to code up any formula what-so-ever. Mostly I
> couldn't care, because the specific numerical values and formulas used are
> irrelevant to the issue of how truth values are handled, from the
> engineering perspective. That is not where the snakes lie.
>
> To change the formula for the simple truth value, just go here:
>
> https://github.com/opencog/opencog/blob/master/opencog/atomspace/SimpleTruthValue.cc#L75
> and code up any formula desired. Changing the formula there does not in
> any way alter any of the issues that I raised. It has no bearing on any of
> them.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/issues/1279#issuecomment-74999992>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by linas** [comment]
Superficially, that seems quite reasonable.  However, there are two deep problems.  Let me start by noting that there is an equally simple variant to what you propose that is not invasive.  Just create a function that lives outside of the atomspace, in the user application code, with the signature:

tv = my_merge_func(tva, tvb, mergestyle)  and then just say:

A.replacetv (my_merge_func(A.tv, B.tv, stronger))

Thus, the only thing an atom has to really, truly support is "replace", and you just provide the replacement, which you computed in some way.   This approach is identical to what you propose, and I claim, from an engineering point, its better.  Here's why:

Currently, when adding an atom to the atomspace, if the atom is brand new, its just added. But if it already exists in the atomspace, then the TV of the atom being added, and the TV of the one that is already there, are merged.  This is one of several pain points. 

This implies that your mergestyle option would have to be added to literally hundreds of functions: there's all sorts of infrastructure in the atom table for passing around atoms during the course of adding them. There are API's for the atomspace (not atomtable) that provide convenience wrappers for the user. There are more API's for persistence.  These include asynchronous reads and writes; queues on which atoms are stacked.   Oh, there are also scheme and python API's, each offering a dozen different ways of working with atoms. You would need to introduce this mergestyle parameter into all of these different places. Debug all of them, Write unit tests for all of them.  It would be a hugely disruptive task just begging for bugs and build breaks to happen.

That's not all. There's more. What happens when two different threads try to update the same tv at the same time?  Or, as mentioned before: two different cogservers are updating the same atom, maybe at different times, communicating via the shared database?

The above is a can of worms. Thus I claim that the much easier solution is this:  you already have your Atom A in hand.  So you know its TV value. You already have, in hand, some other TV value that you want to merge.  The solution: write your own "my_merge_func" as above. Call it.  Take the TV that it returns, and just say A.setTV(myNewTV).   That's it. you are done. No muss, no fuss.

Well, mostly no fuss. There is still the merge function buried deep in the bowels of the atom table. It is still there. I would like to remove it, do a lumpectomy.  That merge function is the one that spiderwebs its way through the hundred-odd API calls that are layered above it.  Get rid of that one nucleus, and the proposed mergestyle flag does not have to metastasize throughout the API.

**Comment by bgoertzel** [comment]
Hi Linas,

At first blush, I like your suggestion pretty well.

It avoids wiring the specifics of particular truth value types into the
Atomspace infrastructure, which is nice...

Of course, we would want to provide a set of standard merge functions that
developers could use, working with the standardly provided TruthValue
types.  We don't want every developer who uses SimpleTruthValue to have to
invent their own revision rule.  But we could provide, e.g. a standard
library or whatever of PLN merge functions embodying various PLN revision
rules for the PLN-supported TruthValue types...

If we made the change you suggest, though, then, addAtom would almost never
be the right thing to do.   If addAtom doesn't do revision, then it's only
the right thing to do when you want to overwrite prior truth values with
newly loaded truth values (which I suggest will almost never be the
case)....  So a couple points come up:

a) existing code that uses addAtom and assumes that it will support
merger/revision, would all need to be changed to use external TV merge
functions

b) for the very common use case of loading Atoms into the Atomspace, we'd
need to make available more than one option (e.g. load-using-overwriting,
load-using-PLN-revision as two different options ...)

Neither of these seems a huge point though.   Since merge is not currently
implemented correctly for SimpleTV, it follows there is no current code
whose current functionality depends sensitively on addAtom doing revision
correctly -- except perhaps code using CountTV, which I believe is mostly
your own code at present...

-- Ben






On Fri, Feb 20, 2015 at 3:10 PM, Linas Vepštas <notifications@github.com>
wrote:

> Superficially, that seems quite reasonable. However, there are two deep
> problems. Let me start by noting that there is an equally simple variant to
> what you propose that is not invasive. Just create a function that lives
> outside of the atomspace, in the user application code, with the signature:
>
> tv = my_merge_func(tva, tvb, mergestyle) and then just say:
>
> A.replacetv (my_merge_func(A.tv, B.tv, stronger))
>
> Thus, the only thing an atom has to really, truly support is "replace",
> and you just provide the replacement, which you computed in some way. This
> approach is identical to what you propose, and I claim, from an engineering
> point, its better. Here's why:
>
> Currently, when adding an atom to the atomspace, if the atom is brand new,
> its just added. But if it already exists in the atomspace, then the TV of
> the atom being added, and the TV of the one that is already there, are
> merged. This is one of several pain points.
>
> This implies that your mergestyle option would have to be added to
> literally hundreds of functions: there's all sorts of infrastructure in the
> atom table for passing around atoms during the course of adding them. There
> are API's for the atomspace (not atomtable) that provide convenience
> wrappers for the user. There are more API's for persistence. These include
> asynchronous reads and writes; queues on which atoms are stacked. Oh, there
> are also scheme and python API's, each offering a dozen different ways of
> working with atoms. You would need to introduce this mergestyle parameter
> into all of these different places. Debug all of them, Write unit tests for
> all of them. It would be a hugely disruptive task just begging for bugs and
> build breaks to happen.
>
> That's not all. There's more. What happens when two different threads try
> to update the same tv at the same time? Or, as mentioned before: two
> different cogservers are updating the same atom, maybe at different times,
> communicating via the shared database?
>
> The above is a can of worms. Thus I claim that the much easier solution is
> this: you already have your Atom A in hand. So you know its TV value. You
> already have, in hand, some other TV value that you want to merge. The
> solution: write your own "my_merge_func" as above. Call it. Take the TV
> that it returns, and just say A.setTV(myNewTV). That's it. you are done. No
> muss, no fuss.
>
> Well, mostly no fuss. There is still the merge function buried deep in the
> bowels of the atom table. It is still there. I would like to remove it, do
> a lumpectomy. That merge function is the one that spiderwebs its way
> through the hundred-odd API calls that are layered above it. Get rid of
> that one nucleus, and the proposed mergestyle flag does not have to
> metastasize throughout the API.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/issues/1279#issuecomment-75198542>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by linas** [comment]
Well, how about this:  Since its clear that the real merge problem is the merge in the atomspace, not the merge in the class TV, we can do this:

-- Add new merge methods and/or a mergestyle flag to class TruthValue and/or its derived classes (SimpleTruthValue, etc.)  That way, commonly-used merge code can be "easily found".  This will require changes to the scheme and python API's,  but this seems reasonable.

-- Remove the AtomSpace addAtom() method, create two new methods: getAtom()  and newAtom(). (ditto for addNode, addLink methods)  One possible semantics could be:

getNode() only takes type and node name, but not TV.  It returns existing atom, else returns null.
newNode() can take type, name and TV.  Throws error if atom already exists.

The above semantics is kind-of icky, though -- its cumbersome to use.  So instead, how about this:  getNode() and newOrExistingNode().  Neither of these take a TV.  getNode() returns NULL if the atom does not exist.  newOrExistingNode() returns an existing node (with, obviously, its existing TV), if it exsits, else creates a new one, and gives the new one a default TV (possibly the special Null TV??  or default false? true ??)

I like this second variant better, because I think its easier for programmers to use.  The will almost never use getAtom(), they will almost always use newOrExistingAtom(), and then tweak the TV after they get their hands on it.  There is no need to tell the atomspace that the TV has changed.  Atomspace doesn't care.

Now, defacto, newOrExistingNode()  is almost identical to the current addNode()  except that newOrExistingNode() doesn't take a TV.  So: we could just keep addNode() as is, and remove TV from its argument list.    I bet that will cause very minimal changes to existing code.  Mostly the unit tests have to change, probably.

This last is the best I can think of.

**Comment by linas** [comment]
I have started looking at what it takes to remove the TV argument from addNode(). It is becoming very clear that programmers are deeply confused about this, as they are commonly specifying crazy values for it.  I suspect that none of them understand the "merge" semantics, given what I see in the code.

**Comment by linas** [comment]
For some goofy reason, I decided to go ahead and do the above.  It all went smoothly, except for one case: the DimensionalEmbedding unit test no longer passes. I spent a fair amount of time trying to figure out why, as I tried hard to reproduce the semantics of TV merging in the dim embed code.  -- it should have worked, there's only three places where it inserts atoms into the atomspace, and so I was careful to merge TV's in those spots.  No luck. The code is impenetrable and undocumented; I could not follow the logic of what it was doing.

The code is sitting in a branch called "tv" -- I'd like to merge it tomorrow...

To track work on the rest of this, you might want to open a new bug report: describe the formulas to be used, and the idea for the mergstyle flag.

**Comment by linas** [comment]
Oh, one last remark: I beleive the old behavior, of merging done deep in the atomspace, can be regained with one lines of code. So:

Old code was:
```
atomspace->addNode(CONCEPT_NODE, "blah", tv);
```
equivalent new code:
```
Handle h = atomspace->addNode(CONCEPT_NODE, "blah");
h->merge(tv);
```
which can be shortened to one line:
```
atomspace->addNode(CONCEPT_NODE, "blah")->merge(tv);
```

**Comment by bgoertzel** [comment]
That's very elegant, thanks!

This should be documented on the wiki somewhere, right?

On Sat, Feb 21, 2015 at 1:54 PM, Linas Vepštas <notifications@github.com>
wrote:

> Oh, one last remark: I beleive the old behavior, of merging done deep in
> the atomspace, can be regained with one lines of code. So:
>
> Old code was:
>
> atomspace->addNode(CONCEPT_NODE, "blah", tv);
>
> equivalent new code:
>
> Handle h = atomspace->addNode(CONCEPT_NODE, "blah");
> h->merge(tv);
>
> which can be shortened to one line:
>
> atomspace->addNode(CONCEPT_NODE, "blah")->merge(tv);
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/issues/1279#issuecomment-75357647>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by bgoertzel** [comment]
On Sat, Feb 21, 2015 at 7:04 AM, Linas Vepštas <notifications@github.com>
wrote:

> I have started looking at what it takes to remove the TV argument from
> addNode(). It is becoming very clear that programmers are deeply confused
> about this, as they are commonly specifying crazy values for it. I suspect
> that none of them understand the "merge" semantics, given what I see in the
> code.
>

Yes, I noticed that recently when walking through some code with the HK
team....  This is part of what motivated me to look into the "merge"
methods...

I need to get everyone to tweak their Atom-creating code to create
meaningful truth values, otherwise using PLN will just result in utter
noise...

**Comment by bgoertzel** [comment]
On Sat, Feb 21, 2015 at 1:34 PM, Linas Vepštas <notifications@github.com>
wrote:

> For some goofy reason, I decided to go ahead and do the above. It all went
> smoothly,
>

Awesome, thanks!


> except for one case: the DimensionalEmbedding unit test no longer passes.
> I spent a fair amount of time trying to figure out why, as I tried hard to
> reproduce the semantics of TV merging in the dim embed code. -- it should
> have worked, there's only three places where it inserts atoms into the
> atomspace, and so I was careful to merge TV's in those spots. No luck. The
> code is impenetrable and undocumented; I could not follow the logic of what
> it was doing.
>

Hmmm...  that code did do useful stuff, though, last I checked...


> The code is sitting in a branch called "tv" -- I'd like to merge it
> tomorrow...
>

Sure....  As the prior merge behavior was not correct anyway for SimpleTV,
this shouldn't screw up anyone's code that uses SimpleTV in any serious
way...


> To track work on the rest of this, you might want to open a new bug
> report: describe the formulas to be used, and the idea for the mergstyle
> flag.
>
ok, sure...

**Comment by bgoertzel** [comment]
Just to be clear for others: Linas's code change provided a framework improvement for handling truth value merging (revision), but the specific truth value formula change I suggested in the original issue remains to be implemented...

**Comment by linas** [comment]
I'm done.  All code merged. Fixed bugs in the Dim embedding module.  The bug was this: the module listens for atom-add/atom-remove signals, but did not listen to tv-changed signals.  After adding listeners for the tv-changed signal, the dim embed module works again.  Anyway .. all unit tests now work.  I'm not planning on doing anything more here.

**Comment by linas** [comment]
Re: elegance: Large parts of the code work with atoms n a very klunky way; many programmers seem unaware that there is a simpler way.   Thus, I see oceans of this style:
```
Handle hSomeReallyLongName = as.addNode(CONCEPT_NODE, "stuff");
Handle hSomeReallySimilarName = as.addNode(CONCEPT_NODE, "more stuff");

std::vector<Handle> handleset;
handleset.push_back(hSomeReallyLongName);
handleset.push_back(hSomeReallySimilarName);
Handle hSomeNameForListLink = as.addLink(LIST_LINK, handleset);

// Do it again for PredicateNode and EvaluationLink
```

The non-klunky way to do this is:
```
as.addLink(LIST_LINK,
   as.addNode(CONCEPT_NODE, "stuff"),
   as.addNode(CONCEPT_NODE, "more stuff")
);
```
which is three lines of readable code instead of 3 unreadable lines.  One can even do better: locally, just write
```
#define an as.addNode
#define al as.addLink
```
 then
```
al(LIST_LINK,
   an(CONCEPT_NODE, "stuff"),
   an(CONCEPT_NODE, "more stuff")
);
```
if you need to set TV's:
```
al(LIST_LINK,
   an(CONCEPT_NODE, "stuff")->hmerge(tva),
   an(CONCEPT_NODE, "more stuff")->hmerge(tvb)
)->hmerge(tvc);
```

hmerge() is just like merge, except that it returns a handle, so you can keep working with handles.

**Comment by inflector** [comment]
Thanks for the above @linas that makes things much easier than they appeared based on my reading other people's code.

**Comment by bgoertzel** [comment]
Cool, thanks!!

On Sun, Feb 22, 2015 at 3:01 AM, Linas Vepštas <notifications@github.com>
wrote:

> I'm done. All code merged. Fixed bugs in the Dim embedding module. The bug
> was this: the module listens for atom-add/atom-remove signals, but did not
> listen to tv-changed signals. After adding listeners for the tv-changed
> signal, the dim embed module works again. Anyway .. all unit tests now
> work. I'm not planning on doing anything more here.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/issues/1279#issuecomment-75386705>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by bgoertzel** [comment]
Looks like more good material for the wiki site ;p ;)

On Sun, Feb 22, 2015 at 3:18 AM, Linas Vepštas <notifications@github.com>
wrote:

> Re: elegance: Large parts of the code work with atoms n a very klunky way;
> many programmers seem unaware that there is a simpler way. Thus, I see
> oceans of this style:
>
> Handle hSomeReallyLongName = as.addNode(CONCEPT_NODE, "stuff");
> Handle hSomeReallySimilarName = as.addNode(CONCEPT_NODE, "more stuff");
>
> std::vector<Handle> handleset;
> handleset.push_back(hSomeReallyLongName);
> handleset.push_back(hSomeReallySimilarName);
> Handle hSomeNameForListLink = as.addLink(LIST_LINK, handleset);
>
> // Do it again for PredicateNode and EvaluationLink
>
> The non-klunky way to do this is:
>
> as.addLink(LIST_LINK,
>    as.addNode(CONCEPT_NODE, "stuff"),
>    as.addNode(CONCEPT_NODE, "more stuff")
> );
>
> which is three lines of readable code instead of 3 unreadable lines. One
> can even do better: locally, just write
>
> #define an as.addNode
> #define al as.addLink
>
> then
>
> al(LIST_LINK,
>    an(CONCEPT_NODE, "stuff"),
>    an(CONCEPT_NODE, "more stuff")
> );
>
> if you need to set TV's:
>
> al(LIST_LINK,
>    an(CONCEPT_NODE, "stuff")->hmerge(tva),
>    an(CONCEPT_NODE, "more stuff")->hmerge(tvb)
> )->hmerge(tvc);
>
> hmerge() is just like merge, except that it returns a handle, so you can
> keep working with handles.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/issues/1279#issuecomment-75387631>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by AmeBel** [comment]
moved a [sheet from google doc] (https://docs.google.com/spreadsheets/d/1iSXwM2qSBuIXqRXzrORYM-eiI0mdUuiIE0RH3J5VTzw/edit#gid=234367133) to the [wiki](http://wiki.opencog.org/w/Truth_Value_Types_of_Atoms)

**Comment by AmeBel** [comment]
With respect to @linas  4th and 5th question on https://github.com/opencog/opencog/issues/1279#issuecomment-74942803 above

For a mindagent like PLN (or a user) how can the type of TV be specified? For example, if user x wan't to experiment with ConceptNodes which have Fuzzy TV(or log(probability TV for that matter), then how will he/she be able to make PLN or other mindagents be aware of it?

**Comment by bgoertzel** [comment]
Each Atom has a TruthValue object associated with it ... so when the Atom
is created, a TV is created and associated with it...

Calling getTruthValue() on an Atom gets the TV object, and once any
MindAgent or whatever has the TV object in hand, it can check what type it
is, right?



On Sun, Feb 22, 2015 at 9:15 PM, Amen Belayneh <notifications@github.com>
wrote:

> With respect to @linas <https://github.com/linas> 4th and 5th question on #1279
> (comment)
> <https://github.com/opencog/opencog/issues/1279#issuecomment-74942803>
> above
>
> For a mindagent like PLN (or a user) how can the type of TV be specified?
> For example, if user x wan't to experiment with ConceptNodes which have
> Fuzzy TV(or log(probability TV for that matter), then how will he/she be
> able to make PLN or other mindagents be aware of it?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/issues/1279#issuecomment-75434893>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by AmeBel** [comment]
Yes, you are right, getType() will return TruthValueType for the TV object. Thus your suggestion of splitting SimpleTruthValue to SimpleProbabilisticTV and SimpleFuzzyTV should solve the issue with Fuzzy/Probabilistic TV.

**Comment by cosmoharrigan** [comment]
@AmeBel Thanks for adding that list to the wiki from the spreadsheet. Do you think the 'Rules' tab of the spreadsheet should be there too?

**Comment by AmeBel** [comment]
i am starting the porting of rules to the new rule engine.  For that purpose i thinking of adding the SimpleProbabilisticTV and SimpleFuzzyTV first. Once the porting is done, the documentation in the PLN rulebase will have the details in the 'Rules' tab.

**Comment by AmeBel** [comment]
oh @linas has already added the TVs

**Comment by williampma** [comment]
Are the new ProbablisticTruthValue & FuzzyTruthValue supposed to completely replace SimpleTruthValue, or coexist side by side SimpleTruthValue like they do now?

It even sounds like they should be inheriting from SimpleTruthValue?

**Comment by bgoertzel** [comment]
They should inherit from SimpleTV

On Tue, Feb 24, 2015 at 3:16 PM, William Ma <notifications@github.com>
wrote:

> Are the new ProbablisticTruthValue & FuzzyTruthValue supposed to
> completely replace SimpleTruthValue, or coexist side by side
> SimpleTruthValue like they do now?
>
> It even sounds like they should be inheriting from SimpleTruthValue?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/issues/1279#issuecomment-75710110>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by linas** [comment]
I assumed that they would co-exist, and get used on an as-needed basis. 

You probably should not make them inherit from SimpleTV  There are several  reasons for C++ inheritance first: to define an API, second, to simplify the code.   Part 1 is already obtained by inheriting from TruthValue  and nothing will get simpler by inheriting from SimpleTV. ... it willl only get more complicated.

**Comment by linas** [comment]
Disagree with Ben. Inheriting from SimpleTV will only make the code more complex, harder to read, debug, understand.   Lets not do that  Lets keep things as  simple as possible.

**Comment by bgoertzel** [comment]
Hmmm....  Well, conceptually they inherit from SimpleTV.   As a software
design choice, since this is very simple code anyway, I don't have a strong
opinion....

The reason to make them inherit from SimpleTV would be, for example, that
they both should use the same merge method.   Of course this can be
achieved other ways than C++ inheritance, though...

On Tue, Feb 24, 2015 at 3:27 PM, Linas Vepštas <notifications@github.com>
wrote:

> Disagree with Ben. Inheriting from SimpleTV will only make the code more
> complex, harder to read, debug, understand. Lets not do that Lets keep
> things as simple as possible.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/issues/1279#issuecomment-75710979>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by linas** [comment]
I'm confused,... if they both use the same merge method, then what's the point? why have two different TV's, if they both work exactly the same way?

The only reason for creating Probabilistic vs, Fuzzy is that presumably, the have very different merge functions. If that's not so, then I don't understand how they could possibly be different ...

**Comment by bgoertzel** [comment]
The default merge rule for SimpleProbabilisticTV and FuzzyTV would be the
same.   However, there are also fancier merge rules that can be used for
SimpleProbabilisticTV, but only sometimes... (depending on what other info
is available in the Atomspace).

Also, other rules may treat the two TV types differently.  For example
AND-ing two probabilistic TVs is handled by a different formula than
AND-ing two FuzzyTVs ...



On Tue, Feb 24, 2015 at 3:38 PM, Linas Vepštas <notifications@github.com>
wrote:

> I'm confused,... if they both use the same merge method, then what's the
> point? why have two different TV's, if they both work exactly the same way?
>
> The only reason for creating Probabilistic vs, Fuzzy is that presumably,
> the have very different merge functions. If that's not so, then I don't
> understand how they could possibly be different ...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/issues/1279#issuecomment-75711940>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by linas** [comment]
Ah.. huh. OK. Well, sure, then both can inherit from SimpleTV.

**Comment by williampma** [comment]
Can SimpleTV exist without being either ProbabilisitcTV or FuzzyTV?

**Comment by AmeBel** [comment]
why is there a count in probabilisticTV and not in FuzzyTV?

**Comment by bgoertzel** [comment]
no good reason... actually for revising two fuzzy TVs count should also be
useful IMO

On Wed, Feb 25, 2015 at 11:14 AM, Amen Belayneh <notifications@github.com>
wrote:

> why is there a count in probabilisticTV and not in FuzzyTV?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/issues/1279#issuecomment-75898207>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by misgeatgit** [comment]
n = n1 + n2 - c min(n1,n2)
...
The STV uses (s,c) where c = n /(n+k), where n is the count...
>@bgoertzel  so is this an update formula for c given c was set to an initial value?

**Comment by bgoertzel** [comment]
Those are different c's, not the same parameter...

On Mon, Mar 2, 2015 at 5:17 PM, Misgana Bayetta <notifications@github.com>
wrote:

> n = n1 + n2 - c min(n1,n2)
> ...
> The STV uses (s,c) where c = n /(n+k), where n is the count...
>
> @bgoertzel <https://github.com/bgoertzel> so is this an update formula
> for c given c was set to an initial value?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencog/opencog/issues/1279#issuecomment-76679556>.
>



-- 
Ben Goertzel, PhD
http://goertzel.org

"The reasonable man adapts himself to the world: the unreasonable one
persists in trying to adapt the world to himself. Therefore all progress
depends on the unreasonable man." -- George Bernard Shaw

**Comment by misgeatgit** [comment]
This issue can be closed now since it is addressed by https://github.com/opencog/opencog/pull/1336

**Comment by inflector** [comment]
Any chance this is causing the recent failure of TVMergeUTest:

http://buildbot.opencog.org:8010/builders/opencog_master_trusty/builds/270/steps/test/logs/stdio

This just started failing on the buildbot today.

**Comment by misgeatgit** [comment]
@inflector thanks,I am taking a look at it.

**Comment by misgeatgit** [comment]
@bgoertzel I think you can close this issue now.

**Comment by linas** [comment]
I just closed this, I assume everything is OK now. If there are still issues, open a new bug report ...
